crates/forge-runner/src/running.rs

```diff
@@ -388,3 +395,59 @@ fn get_call_trace_ref(runtime: &mut ForgeRuntime) -> Rc<RefCell<CallTrace>> {
         .current_call_stack
         .top()
 }
+
+fn add_header(
+    entry_code: Vec<Instruction>,
+    assembled_program: &mut AssembledCairoProgramWithSerde,
+) {
+    let mut new_bytecode = vec![];
+    let mut new_hints = vec![];
+    for instruction in entry_code {
+        if !instruction.hints.is_empty() {
+            new_hints.push((new_bytecode.len(), instruction.hints.clone()));
+        }
+        new_bytecode.extend(instruction.assemble().encode().into_iter());
+    }
+
+    let new_bytecode_len = new_bytecode.len();
+    assembled_program.hints.iter_mut().for_each(|hint| {
+        hint.0 += new_bytecode_len;
+    });
+
+    assembled_program.hints = [new_hints, assembled_program.hints.clone()].concat();
+    assembled_program.bytecode = [new_bytecode, assembled_program.bytecode.clone()].concat();
+}
+
+fn add_footer(footer: Vec<Instruction>, assembled_program: &mut AssembledCairoProgramWithSerde) {
+    for instruction in footer {
+        assembled_program
+            .bytecode
+            .extend(instruction.assemble().encode().into_iter());
+    }
+}
+
+fn create_hints_dict(
+    assembled_program: &mut AssembledCairoProgramWithSerde,
+) -> (HashMap<String, Hint>, HashMap<usize, Vec<HintParams>>) {
+    let string_to_hint: HashMap<String, Hint> = assembled_program
+        .hints
+        .iter()
+        .flat_map(|(_, hints)| hints.iter().cloned())
+        .map(|hint| (hint.representing_string(), hint))
+        .collect();
+    let hints_dict = assembled_program
+        .hints
+        .iter()
+        .map(|(offset, hints)| {
+            (
+                *offset,
+                hints
+                    .iter()
+                    .map(hint_to_hint_params)
+                    .collect::<Vec<HintParams>>(),
+            )
+        })
+        .collect();
+
+    (string_to_hint, hints_dict)
+}
```

Comment:
Are those copied from the compiler but use our structs?

Reply:
Not entirely. Code in those methods is inspired but have some fundamental differencies